---
title: "Differential expression analysis using DEseq2"
author: "David Requena and Jamie Saltsman"
date: "10/04/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Differential expression analysis using DEseq2

## 1. SET UP

### Install and/or call the required libraries

This first block is just to install the dependencies. If already installed, just left this commented.

```{r Install libraries, eval=FALSE, echo=FALSE}
# # Import necessary libraries
# install.packages('knitr')
# install.packages('matrixStats')
# install.packages('RColorBrewer')
# install.packages('gplots')
# install.packages('ggplot2')
# install.packages('ggrepel')
# install.packages('tsne')
# install.packages('BatchJobs')
# install.packages('gridExtra')
 
#source("https://bioconductor.org/biocLite.R")
# biocLite("XML") #In linux, first do: sudo apt-get install libxml2-dev
# biocLite("BiocParallel", dependencies=TRUE)
# biocLite("DESeq2", dependencies=TRUE)
# biocLite("biomaRt", dependencies=TRUE)
# biocLite("pheatmap", dependencies=TRUE)
```
This block is to call all the required libraries:

```{r, echo=FALSE}
# library("XML")
library("knitr")
library("DESeq2")
library("biomaRt")
library("matrixStats")
library("BatchJobs")
library("BiocParallel")
library("RColorBrewer")
library("gplots")
library("ggplot2")
library("ggrepel")
library("gridExtra")
library("tsne")
library("pheatmap")
library("grDevices")
```

### Input sample metadata

In this block, you just need to change the working directory:

```{r "setup"}
# require("knitr")
knitr::opts_knit$set(root.dir = "../RNAseq_data") # <<<------------- *** WORKING DIRECTORY ***

# These two tables have to be in your working directory:
# - A table containing the samples' data (features of interest) and
# - A table with the gene counts by sample
sample_data <- "samples_data.tsv"
counts_data <- "counts_data.tsv"
```

This block is to read the samples' data.
There is also an option to work with subsets.

```{r, echo=FALSE}
#########################
# Input sample metadata #
#########################

sampledata <- data.frame(read.csv(sample_data, sep = "\t", header = TRUE))

# The next three lines are examples of how to work with a subset of the data:
#sampledata <- sampledata[sampledata$library_prep!=0,]
#sampledata <- sampledata[sampledata$have_npm=="Y",]
#sampledata <- sampledata[sampledata$normal_pm=="Y",]

sampledata$library_prep <- factor(sampledata$library_prep)
droplevels(sampledata$id)

# To sort the table by sample name:
sampledata <- sampledata[order(sampledata[,2]),]
sampledata$patient <- as.character(sampledata$patient)
rownames(sampledata) <- sampledata$id
```

This block is to read the tables with the counts.
There is also an option to filter out some genes (blacklist).
```{r, echo=FALSE}
######################
# Input counts table #
######################

countstable <- data.frame(read.csv(counts_data, sep = "\t", header = FALSE), stringsAsFactors = FALSE)

# Filter out some genes in a pre defined black list:
blacklist <- data.frame(read.csv("blacklist.tsv", sep = "\t", header = TRUE))
countstable <- countstable[!(countstable[,1] %in% blacklist[,1]), ]

# Save the gene list, to be used later as row names:
generows <- unlist(countstable[,1], use.names=FALSE)
generows <- generows[-1]
invisible(droplevels(generows))

# This column contains the gene names, we don't need it anymore:
countstable[,1] <- NULL

# Sort the table by sample name:
countstable.sorted <- countstable[,order(countstable[1,])]

# Save the samples list, to be used later as column names:
genecols <- unlist(countstable.sorted[1,], use.names=FALSE)
droplevels(genecols)

# Convert the values on the table from characters to numbers:
# First, any string should be removed (names on the first row)
countstable.sorted <- countstable.sorted[-1,]
# Then, do the conversion
countstable.num <- as.data.frame(apply(countstable.sorted, c(1,2), as.integer))
# Finally, add the row and column names:
rownames(countstable.num) <- generows
colnames(countstable.num) <- genecols

# If the sampledata table were filtered in the chunk above,
# this line is to filter the countstable
# If not, this line will do nothing
countstable.num <- countstable.num[colnames(countstable.num) %in% sampledata$id]
```
This block is to explore the distribution of counts by sample
```{r, echo=FALSE}
##################################################
# Exploring the distribution of counts by sample #
##################################################

# Histogram of the samples with zero counts
h <- hist(rowSums(countstable.num == 0), breaks = seq(-0.5, ncol(countstable.num) + 0.5, 1), labels = TRUE)

# To inspect the values
h

# To plot the histogram of the total number of samples
plot(h, xaxp = c(0, ncol(countstable.num), ncol(countstable.num)), labels = TRUE,
     main = paste0("Histogram of samples with no counts, by gene\n(Total number of samples = ",
                   ncol(countstable.num), ")"),
     xlab = "Number of samples with no counts", ylab = "Number of genes")
abline(v = ncol(countstable.num) - 0.5, col = "red", lty = 2)
abline(v = ncol(countstable.num) - 1.5, col = "orange", lty = 2)
text(ncol(countstable.num) - 0.1, tail(h$counts, 1) * 1.2, col = "red", srt = 60,
     label = paste0(round(tail(h$counts, 1) * 100 / nrow(countstable.num), digits = 2), "%"))
text(ncol(countstable.num) - 1.1, tail(h$counts, 1) * 1.15, col = "orange", srt = 60,
     label = paste0(round(sum(tail(h$counts, 2)) * 100 / nrow(countstable.num), digits = 2), "%"))
```

Prior Filtering
```{r, echo=FALSE}

# Remove empty rows:
countstable.num <- countstable.num[ rowSums(countstable.num) > 0, ]

# You should remove also those genes with counts in only one sample.
# But there have to be a rationale for this. Check first those cases,
# to determine if this is reasonable to drop them or not.
# To check the counts in those samples, uncomment the next two lines:
# genes_1sample <- countstable.num[rowSums(countstable.num > 0) == 1, ]
# genes_1sample$sum <-rowSums(genes_1sample)

# To get their names and functions, uncomment the next 9 lines:
# ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
# genemap <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "description"),
#                  filters = "ensembl_gene_id",
#                  values = rownames(genes_1sample),
#                  mart = ensembl)
# 
# indx <- match(rownames(genes_1sample), genemap$ensembl_gene_id)
# genes_1sample$symbol <- genemap$hgnc_symbol[indx]
# genes_1sample$description <- genemap$description[indx]

# Remove those rows (genes) with only one sample with counts:
countstable.num <- countstable.num[rowSums(countstable.num > 0) > 1, ]

# The rest of the code in this block is only necessary if you wanna apply an additional
# independent prior filtering before the differential expression analysis:

##########################################
# Independent prior filtering (optional) #
##########################################

# Two additional types of prior filtering are provided, as examples:

# Remove any gene with 0 reads in at least one sample (using a different code)
# (i.e. considering only transcripts present in all the samples)
# countstable.num <- cbind(countstable.num, 1)
# for (i in 1:(ncol(countstable.num) - 1)) {
#   countstable.num$`1` <- countstable.num$`1` * countstable.num[,i]
# }
# countstable.num <- countstable.num[countstable.num$`1` != 0,]
# countstable.num$`1` <- NULL

# Remove genes with less than 5 reads in every sample
# countstable.num <- countstable.num[apply(countstable.num, 1, function(row) (all(row > 5))),]
```

### Create the DEseq2 object
With no specific design, but indicating the base level
```{r, echo=FALSE}
################
# DESeq Object #
################

# This is to reate a DESeqDataSet from the tables above, with no specific design yet:
ds.deseq <- DESeqDataSetFromMatrix(
  countData = countstable.num,
  colData = sampledata,
  design = ~ sample_type)

# To see which is the current base level:
str(colData(ds.deseq)$sample_type)
# To indicate the base level: sample_type == normal
colData(ds.deseq)$sample_type  <- relevel(colData(ds.deseq)$sample_type, "normal")
# To verify which is the current base level
str(colData(ds.deseq)$sample_type)
```

## 2. Explore the data
This step is necessary to determine which variables we should control by

### Transformations of the Data
To adjust by known effects
```{r, echo=FALSE}
# Variance stabilized transformation
transf.data <- varianceStabilizingTransformation(ds.deseq)
head(assay(transf.data))

# rLog transformation
# transf.data <- rlog(ds.deseq)
# head(assay(transf.data))
```

### PCA plot
```{r, eval = TRUE}
######################
# plotPCAdr function #
######################
# This is the plotPCA function from DESeq2, made by Wolfgang Huber
# But including some improvements made by David Requena
# Now this allows to choose which PCs plot and to use two features for color/shape

plotPCAdr <- function (object, intgroup = "condition", ntop = 500, PCs = c(1,2), returnData = FALSE) 
{
  PC_x <- paste("PC", PCs[1], sep = "")
  PC_y <- paste("PC", PCs[2], sep = "")
  rv <- rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  
  #return(pca$x)
  #return(pca$x[, c(PCs[1], PCs[2])])
  
  # To assign different shapes
  d <- data.frame(PCx = pca$x[, PCs[1]], PCy = pca$x[, PCs[2]],
                  to_color = factor(intgroup.df[,1]), to_shape = factor(intgroup.df[,2]),
                  name = colnames(object))
  
  if (returnData) {
    attr(d, "percentVar") <- percentVar[PCs[1]:PCs[2]]
    return(d)
  }

  ggplot(data = d, aes_string(x = "PCx", y = "PCy", color = "to_color", shape = "to_shape")) +
    labs(color = intgroup[1], shape = intgroup[2]) + geom_point(size = 3, alpha = 1) +
    xlab(paste0(PC_x, ": ", round(percentVar[PCs[1]] * 100), "% variance")) +
    ylab(paste0(PC_y, ": ", round(percentVar[PCs[2]] * 100), "% variance")) + coord_fixed()
  
  # To use only one feature:
  # Comment 13 lines of the code above [from "d <-" to  "ylab(..."]
  # and uncomment the next 14 lines [from "group <- " to  "ylab(..."]
  # group <- if (length(intgroup) > 1) {
  #   factor(apply(intgroup.df, 1, paste, collapse = " : "))
  # }
  # else {
  #   colData(object)[[intgroup]]
  # }
  #
  # d <- data.frame(PCx = pca$x[, PCs[1]], PCy = pca$x[, PCs[2]], group = group,
  #                 intgroup.df, name = colnames(object))
  #
  # ggplot(data = d, aes_string(x = "PCx", y = "PCy", color = "group")) +
  #   geom_point(size = 3, alpha = 1) + coord_fixed() +
  #   xlab(paste0(PC_x, ": ", round(percentVar[PCs[1]] * 100), "% variance")) +
  #   ylab(paste0(PC_y, ": ", round(percentVar[PCs[2]] * 100), "% variance"))
}

##############
# Make some PCA plots:

# This code is to use a simpler method to plot the PCs:
pr <- prcomp(t(assay(transf.data)))

# This code is to use the function defined above

# PC 1 vs 2
pca12 <- plotPCAdr(transf.data,
                   intgroup = c("sample_type", "lib_prep"),
                   ntop =  nrow(assay(transf.data)),
                   PCs = c(1,2))
pca12 <- pca12 + geom_text_repel(aes_string(label = transf.data$id), color = "black",
                             cex=3, nudge_x = -4, nudge_y = 0,
                             min.segment.length = unit(0.1, "lines"), box.padding = unit(0.4, "lines"))

# PC 2 vs 3
pca23 <- plotPCAdr(transf.data,
                   intgroup = c("sample_type", "lib_prep"),
                   ntop =  nrow(assay(transf.data)),
                   PCs = c(2,3))
pca23 <- pca23 + geom_text_repel(aes_string(label = transf.data$id), color = "black",
                             cex=3, nudge_x = -4, nudge_y = 0,
                             min.segment.length = unit(0.1, "lines"), box.padding = unit(0.4, "lines"))

# PC 3 vs 4
pca34 <- plotPCAdr(transf.data,
                   intgroup = c("sample_type", "lib_prep"),
                   ntop =  nrow(assay(transf.data)),
                   PCs = c(3,4))
pca34 <- pca34 + geom_text_repel(aes_string(label = transf.data$id), color = "black",
                             cex=3, nudge_x = -4, nudge_y = 0,
                             min.segment.length = unit(0.1, "lines"), box.padding = unit(0.4, "lines"))

#############################
# Show the 3 PCA plots made #
#############################

# PC 1 vs 2:
############
# No labels
plot(pr$x[,])
# Labels, colors and shapes
pca12

# PC 2 vs 3:
############
# No labels
plot(pr$x[,-1])
# Labels, colors and shapes
pca23

# PC 3 vs 4:
############
# No labels
plot(pr$x[,-c(1,2)])
# Labels, colors and shapes
pca34

# This is to add labels to the basic PCA plot made using the function plot()
# But this is not necessary because we have now the function plotPCAdr
# text(pr$x[,2], pr$x[,3], labels = colnames(assay(transf.data)),
#      pos = 1, offset = 0.2, cex = 0.7)

# This is to extract the individual gene contribution to PC1
# sort(abs(pr$rotation[,1]), decreasing = TRUE)[1:50]
# sort(abs(pr$rotation[,1]), decreasing = TRUE)
# write.csv(sort(abs(pr$rotation[,1]), decreasing = TRUE), file = "PC1.csv")
```

### tSNE plot
```{r, echo=FALSE}
# Calculate the euclidean distances between samples
sampleDists <- dist(t(assay(transf.data)))

# Perplexity is set to 4 because this is our average number of samples by patient (neighbors in a set)
samples.tsne <- tsne(sampleDists, perplexity = 4, max_iter = 10000, epoch=1000)

df.tsne <- data.frame(samples.tsne)
df.tsne$id <- sampledata$id
df.tsne$library <- sampledata$lib_prep
df.tsne$patient <- sampledata$patient
df.tsne$sample_type <- sampledata$sample_type
df.tsne$sample_type <- factor(df.tsne$sample_type, levels=c("normal", "primary", "metastasis"),
                              labels=c("Normal", "Primary", "Metastasis"))

nice_colors <- c("Normal" = "dodgerblue3", "Primary" = "gold", "Metastasis" = "firebrick3")

# tSNE plot de-identifying the samples:
p.tsne_clean <- ggplot(data = df.tsne) + geom_point(aes(x = X1, y = X2), size = 4) + theme_bw()
p.tsne_clean

# tSNE plot with labels, colors and shapes:
p.tsne <- ggplot(data = df.tsne) + theme_bw() +
  geom_point(aes(x = X1, y = X2, color = sample_type, shape = library), size = 4) +
  geom_text_repel(aes_string(x = df.tsne$X1, y = df.tsne$X2, label = df.tsne$id), color = "black",
                  cex=3, nudge_x = -4, nudge_y = 0,
                  min.segment.length = unit(0.1, "lines"), box.padding = unit(0.4, "lines")) +
  scale_fill_manual(values = nice_colors)

p.tsne

# This is a bigger version of the plot, recommended for posters
# Warning: set appropriate for the legend according to your plot
p.tsne_big <- ggplot(data = df.tsne, aes(x = X1, y = X2)) +
  geom_point(aes(fill = sample_type, shape = library), size = 12, alpha = 0.9) + theme_bw() +
  scale_shape_manual(values = c(21, 22, 23, 24)) + scale_fill_manual(values = nice_colors) +
  geom_text(aes(label= patient), size = 6) +
  theme(axis.text.x = element_blank(), axis.text.y = element_blank(),
        legend.position = c(0.87, 0.70), legend.title = element_text(size=20),
        legend.text=element_text(size=16), legend.background = element_rect(color = "black")) +
  labs(fill = expression(underline("Sample Type")), shape = expression(underline("Library")))
        
p.tsne_big

ggsave("Plots/tSNE_clean.jpg", p.tsne, width = 16, height = 10, dpi = 300)
ggsave("Plots/tSNE.jpg", p.tsne, width = 16, height = 10, dpi = 300)
ggsave("Plots/tSNE_big.jpg", p.tsne, width = 8.5, height = 7, dpi = 300)

# To include text inside the circles:

```

### HeatMap
```{r, echo=FALSE}
sampleDistMatrix <- as.matrix(sampleDists)

rownames(sampleDistMatrix) <- paste(transf.data$id, transf.data$sample_type, sep = "_")
colnames(sampleDistMatrix) <- paste(transf.data$id, transf.data$sample_type, sep = "_")

# Make heatmap based on the Distance matrix (uses all genes)
heatmap.2(sampleDistMatrix, trace="none", margins = c(7.2, 7.2), offsetRow = 0, offsetCol = 0,
          key = TRUE, density.info = "none")

# To personalize the colors, use:
# Colors = colorRampPalette(rev(brewer.pal(9, "Reds"))) (255)
# heatmap.2(sampleDistMatrix, trace="none", margins = c(7.2, 7.2), offsetRow = 0, offsetCol = 0,
#          key = TRUE, density.info = "none", col=colors)

# Using a different function, to cluster also by genes
pheatmap(transf.data, cluster_rows = TRUE, show_rownames = FALSE, #kmeans_k = 1000,
         annotation = data.frame(sampledata[,c(1,6)]))
```


## 3. Data Analysis

### Model matrix
Using the variables chosen from the results above, a model will be determined.
A matrix representing this model is provided to DEseq in this step.
```{r, echo=FALSE}
################
# Model Matrix #
################

# Define the model (design) to use. In our case, our variable of interest is the sample type
# (normal, tumor or metastasis). But, we have to control by:
# library preparation -> lib_prep
# patient -> lib_prep:nested_patient
# Due patient is collinear with library prep, this should be represented by a combination
# of lib_prep and a secondary variable (nested patient)
mm <- model.matrix(~ lib_prep + lib_prep:nested_patient + sample_type, colData(ds.deseq))

# Sometimes, the resulting matrix has columns full of zeros. These should be removed:
mm <- mm[ , colSums(mm) > 0]
```

### Comparison
```{r, echo=FALSE}
# Run DESeq2 using the model defined above
ds.deseq <- DESeq(ds.deseq, full=mm,
                  modelMatrixType="standard", betaPrior=FALSE,
                  parallel = TRUE, BPPARAM=BatchJobsParam(workers = 64))

# This will show the possible comparisons, according to the design provided
resultsNames(ds.deseq)

# Cut off p-value
cutoff_alpha <- 0.05
# Cut off fold-change
cutoff_fold <- 1

# Get the result of a certain comparison (see "help(results)" for info on this)
res <- results(ds.deseq, contrast=list("sample_typemetastasis", "sample_typeprimary"),
               altHypothesis="greaterAbs",
               alpha = cutoff_alpha,
               pAdjustMethod = "BH") # Benjamini Hochberg, or FDR

summary(res)
```

### Annotation and output tables
This block is to add the names and description of the genes.
```{r, echo=FALSE}
#####################
# Adding gene names #
#####################
res$ensembl <- rownames(res)

ensembl = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
# More annotations could also be added from biomaRt:
# listAttributes(mart = ensembl)

genemap <- getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "description"),
                 filters = "ensembl_gene_id",
                 values = res$ensembl,
                 mart = ensembl)

idx <- match(res$ensembl, genemap$ensembl_gene_id)
res$symbol <- genemap$hgnc_symbol[idx]
res$description <- genemap$description[idx]
```

This is to visualize a summary of the results and make some output tables in .csv format
```{r, echo=FALSE}
#########################################
# Visualizing and filtering the results #
#########################################

summary(res)
write.csv(as.data.frame(res), file = "results_Mets_vs_Prims_full.csv")
head(res, n = 10)

table(res$padj < cutoff_alpha)
table(abs(res$log2FoldChange) >= cutoff_fold)
table((abs(res$log2FoldChange) >= cutoff_fold) & (res$padj < cutoff_alpha))

res_sig <- subset(res, ( (padj < cutoff_alpha) & !is.na(padj) ))
summary(res_sig)
head(res_sig, n = 10)
#write.csv(as.data.frame(res), file = "results_Mets_vs_Prims_sig.csv")

res_fold.over <- subset(res_sig, log2FoldChange >= cutoff_fold)
res_fold.under <- subset(res_sig, log2FoldChange <= -cutoff_fold)

res_fold.over.sorted <- res_fold.over[order(-res_fold.over$log2FoldChange, na.last=TRUE), ]
res_fold.under.sorted <- res_fold.under[order(res_fold.under$log2FoldChange, na.last=TRUE), ]
head(res_fold.over.sorted, n = 10)
head(res_fold.under.sorted, n = 10)

write.csv(as.data.frame(res_fold.over.sorted), file = "results_Mets_vs_Prims_over_(pval_filtered)_(fold_sorted).csv")
write.csv(as.data.frame(res_fold.under.sorted), file = "results_Mets_vs_Prims_under_(pval_filtered)_(fold_sorted).csv")
```
### Plots
This code is to generate some plots from the results

#### Volcano Plot
```{r, echo=FALSE}

# New data frame for the volcano plot
d.volcano <- data.frame(res)
# To remove empty adjusted p-values
d.volcano <- d.volcano[!is.na(d.volcano$padj), ]

# To reduce the range of the volcano plot, for posters
#d.volcano <- d.volcano[(abs(d.volcano$log2FoldChange) < 8) & (-log10(d.volcano$padj) < 7.5), ]

# To fill the empty labels
d.volcano$symbol[is.na(d.volcano$symbol) | (d.volcano$symbol == "")] <- "NS"

# Additional cut-offs. The script will only plot the labels of the
# data points passing the "nice" cutoff, for a better visualization
ctf_nice <- cutoff_alpha / 500
fld_nice <- cutoff_fold * 3

# Colors
d.volcano$colors <- rep("other", nrow(d.volcano))
d.volcano$colors[(abs(d.volcano$log2FoldChange) <= cutoff_fold) & (d.volcano$padj >= cutoff_alpha)] <- "worst"
d.volcano$colors[(abs(d.volcano$log2FoldChange) <= cutoff_fold) & (d.volcano$padj <= cutoff_alpha)] <- "middle"
d.volcano$colors[(d.volcano$log2FoldChange >= cutoff_fold) & (d.volcano$padj < cutoff_alpha)] <- "over"
d.volcano$colors[(d.volcano$log2FoldChange <= -cutoff_fold) & (d.volcano$padj < cutoff_alpha)] <- "under"

# Condition to select just a group of genes
cond1 <- !( (abs(d.volcano$log2FoldChange) <= cutoff_fold) | (d.volcano$padj > cutoff_alpha) )
cond2 <- !( (abs(d.volcano$log2FoldChange) <= fld_nice) & (d.volcano$padj > ctf_nice) )
cond <- cond1 & cond2

#############################################
# Volcano Plot using ggplot2 and text repel #
#############################################
p.volcano <- ggplot() + theme_bw() +
  scale_y_continuous(expand=c(0,0), limits = c(0, max(-log10(d.volcano$padj)) + 1) ) +
    geom_vline(xintercept = c(-cutoff_fold, cutoff_fold), colour = "red", alpha = 0.8, linetype = 2) +
  geom_hline(yintercept = c(-log10(cutoff_alpha)), colour = "red", alpha = 0.8, linetype = 2) +
  geom_rect(aes(xmin = -fld_nice, xmax = fld_nice, ymin = 0, ymax = -log10(cutoff_alpha)),
            fill="mistyrose", alpha = 0.4, inherit.aes = FALSE) +
  geom_rect(aes(xmin = -cutoff_fold, xmax = cutoff_fold, ymin = 0, ymax = -log10(ctf_nice)),
            fill="mistyrose", alpha = 0.4, inherit.aes = FALSE) +
  geom_rect(aes(xmin = -fld_nice, xmax = -cutoff_fold, ymin = -log10(cutoff_alpha), ymax = -log10(ctf_nice)),
            fill="lightgoldenrodyellow", alpha = 0.5, inherit.aes = FALSE) +
  geom_rect(aes(xmin = cutoff_fold, xmax = fld_nice, ymin = -log10(cutoff_alpha), ymax = -log10(ctf_nice)),
            fill="lightgoldenrodyellow", alpha = 0.5, inherit.aes = FALSE) +
  geom_point(data = d.volcano, aes(x = d.volcano$log2FoldChange, y = -log10(d.volcano$padj), fill = d.volcano$colors),
             size = 1.5, pch=21, color = "gray10", alpha = 0.7, show.legend = TRUE) +
  scale_fill_manual(name = "Relative expression",
                     values = c("over" = "springgreen",
                                "under" = "deepskyblue",
                                "middle" = "yellow",
                                "worst" = "firebrick1",
                                "other" = "tan1"),
                     breaks = c("over", "under"), labels = c("Overexpressed", "Underexpressed")) +
  guides(fill = guide_legend(override.aes = list(size=8))) + theme(legend.position = c(0.90, 0.92)) +
  geom_text_repel(aes(x = d.volcano$log2FoldChange[cond],
                      y = -log10(d.volcano$padj[cond]),
                      label = d.volcano$symbol[cond]),
                  inherit.aes = FALSE, parse = FALSE, max.iter = 5000, color = "black", cex = 4,
                  nudge_x = 0, nudge_y = 0.1, segment.alpha = 0.5, segment.size = 0.5,
                  arrow = arrow(length = unit(0.005, "npc"), type = "closed"),
                  box.padding = unit(0.1, "lines"), min.segment.length = unit(0.01, "lines")) +
  labs(x = expression("log"[2]*"(Fold Change)"),
       y = expression("-log"[10]*"(p adjusted)"),
       title = "Metastasis / Primary tumors")

p.volcano

############################
# Traditional Volcano Plot #
############################
# plot(res$log2FoldChange, -log10(res$padj), pch = 16, cex = 0.5, las = 1,
#      xlab = "log2(Fold Change)", ylab = "-log10(p adjusted)")
# 
# # To add labels (names of the genes):
# text(x = res$log2FoldChange[cond], y = -log10(res$padj[cond]) + 0.1,
#      labels = res$symbol[cond], cex = 0.7)
# 
# # To draw vertical and horizontal lines
# abline(h = -log10(0.05), col = "green")
# abline(v = c(-1, 1), col = "green")
# abline(h = -log10(0.0001), col = "grey35")
# abline(v = c(-3, 3), col = "grey35")
# 
# # To add a title
# title(main = "Metastasis vs Primary")
```

#### HeatMap
```{r, echo=FALSE}
###########
# HeatMap #
###########
# To cluster using a subset of the genes
# In this case, using only the differentially expressed genes passing the cut-off (adjusted p-value)
# And, coloring by patient and sample type
transf.data.sig <- transf.data[rownames(transf.data) %in% rownames(res_sig),]
pheatmap(transf.data.sig, cluster_rows = TRUE, show_rownames = FALSE, #kmeans_k = 100,
         annotation = data.frame(sampledata[,c(1,6)]))
```

#### BoxPlot and ScatterPlot
After the differential expression analysis, the genes of interest can be further analyzed individually using
boxplot and scatterplots.

```{r, echo=FALSE}
###########
# BoxPlot #
###########

# This function will plot a Boxplot and also the datapoints
plotDR_bygene <- function (genename = NULL, symbol = NULL)
{
  # Extracting the vector of counts for that gene
  gene_counts <- counts(ds.deseq, normalized = TRUE)[genename,]
  
  # Making a dataframe for the plot
  df.box <- data.frame(ds.deseq@colData[, c("id", "lib_prep", "sample_type", "patient")], gene_counts)
  
  # Re-ordering sample_type for the plot
  df.box$sample_type <- factor(df.box$sample_type, levels = c("normal", "primary", "metastasis"), labels=c("N", "P", "M"))
  # This combines lib_prep and sample_type in a new variable
  df.box$lib_sample <- interaction(df.box$lib_prep, df.box$sample_type)
  
  # Plot
  p.bs <- ggplot(df.box, aes(x = df.box$sample_type, y = log2(df.box$gene_counts))) +
    theme_bw() + geom_boxplot(width = 0.8, fill = "gray90") +
    geom_point(aes(fill = df.box$sample_type),
               size = 3, alpha = 0.9, pch = 21, color = "black",
               position = position_jitter(width = 0.35)) +
    labs(title = paste0("Gene: ", symbol, " (", genename, ")")) +
    labs(x = expression("Sample Type"),
         y = expression("log"[2]*"(Normalized Gene Counts)"),
         title = paste0("Gene: ", symbol, " (", genename, ")")) +
    scale_fill_discrete(name = "Sample Type")
    
    # Comment 5 lines above and use the following 5 lines instead to make bigger plots, good for posters:
    # labs(title = paste0(symbol, " (", genename, ")")) +
    # theme(axis.title.x = element_blank(), axis.title.y = element_blank(),
    #       axis.text.x = element_text(size=14), axis.text.y = element_text(size=14)) +
    # guides(fill=FALSE) +
    # scale_fill_manual(values = c("N" = "dodgerblue3", "P" = "gold", "M" = "firebrick3"))

  return(p.bs)
}

# The genes of interest will be provided in a .tsv table:
interesting_genes <- data.frame(read.csv("genes_overexp.tsv", sep = "\t", header = TRUE))
interesting_genes <- data.frame(read.csv("genes_underexp.tsv", sep = "\t", header = TRUE))

# Then, this loop will make and save the plots for all the genes in the table
for (i in 1:nrow(interesting_genes)) {
  ggsave(paste0("By_Gene/Underexp/", interesting_genes[i, 2], ".jpg"),
         plotDR_bygene(toString(interesting_genes[i, 1]), toString(interesting_genes[i, 2])),
         width = 16, height = 10, dpi = 300)
}

#aes(shape = df.box$lib_prep)
```

For further guidance, check the DESeq2 Vignette:
https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html
